compiler/rustc_ty_utils/src/opaque_types.rs

```diff
@@ -240,6 +240,10 @@ impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for OpaqueTypeCollector<'tcx> {
                                 continue;
                             }
 
+                            if !self.seen.insert(assoc.def_id.expect_local()) {
+                                return;
+                            }
```

Comment:
one thought: if we first encounter `<Self as Trait>::Gat<u32>` which normalizes to `<u32 as Trait>::Atpit` and then later encounter `<Self as Trait>::Gat<Self>`, which normalizes to `<Self as Trait>::Atpit`, this change causes us to not define `Atpit` anymore, does it?

While it seems fine to not support that for now, I would like to have an open issue/test for that

Reply:
oof right, we had this issue with structs before -.- I knew why I didn't want this check. Maybe I should move the recursion check even later